### DISCUSSION TITLE
Icons in Table will flicker

### DISCUSSION THREAD

**Discussion by junmaqiang** [fyne-io/fyne#3680] [Q&A] [answered]
**
When the column components are inconsistent, the mouse over will shake
**

```
	table := widget.NewTable(
		func() (int, int) {
			return len(a.DeviceData), len(a.DeviceData[0])
		},
		func() fyne.CanvasObject {
			return container.NewMax(widget.NewLabel(""))
		},
		func(id widget.TableCellID, cell fyne.CanvasObject) {
                     //Put a few more button components here, align them under Settings, and the problem arises
                  }
         )
```

**Comment by andydotxyz** [comment]
The code you provide is not sufficient to reproduce what you describe. Is this a discussion? Can you provide some more information?

**Comment by andydotxyz** [comment]
Updating to "Flicker" instead of "Shake" as the positions seem consistent

**Comment by andydotxyz** [comment]
On every re-draw you are asking the label and icon to hide, and then showing one of them again.
That seems like it could cause a flicker - see if you can update your code to not hide things you want to stay visible.

**Reply by junmaqiang** [reply]
Alignment also causes flickering

**Reply by junmaqiang** [reply]
Especially right alignment

**Reply by junmaqiang** [reply]
I think hide or show, align left or align right are general requirements and should not flash

**Reply by Jacalz** [reply]
I'd say that it is logically impossible to not get flickering if you are hiding/showing stuff within a rapid succession. There is not anything we can do about that as far as I can see

**Reply by andydotxyz** [reply]
I agree these are general requirements, and in almost all the apps created it does not flicker. If you can replicate the issue without toggling the wrong position and then right (or hiding then showing) we can look into it.

However as @Jacalz says your code is currently asking for a flicker by resetting state before setting it again. Fyne will attempt to draw when requested - so your `Hide()` call is requesting an item to appear hidden.

**Comment by andydotxyz** [comment]
You should not create new widgets in the "Update" callback.
Doing this will result in a very slow performance and lots of flicker, because creating graphical widgets is not fast.
The `Update` (third) should be used to configure widgets made in the `Create` (second) callback.

Also if you do change fields like `Objects` your code must call `Refresh` for the new configuration to be rendered immediately.

**Reply by junmaqiang** [reply]
Can you use my code for an example reference?  Please

**Comment by junmaqiang** [comment]
### There are too few examples of table components
https://developer.fyne.io/explore/

**Comment by junmaqiang** [comment, marked as answer]
### I have solved the problem
cell.Refresh()  
![截屏2023-02-28 11 49 43](https://user-images.githubusercontent.com/1819457/221748951-aa66312f-15fb-40d0-8ffe-70447d94dc7d.png)
